File: _includes/sort-benchmark/algorithms/bucket.rs
fn bucket_sort(a: &mut [usize]) {
    if a.is_empty() {
        return;
    }

    let n = a.len();
    let min = *a.iter().min().unwrap();
    let max = *a.iter().max().unwrap();
    let bucket_count = n;
    let mut buckets: Vec<Vec<usize>> = vec![Vec::new(); bucket_count];

    for &x in a.iter() {
        let idx = if max == min {
            0
        } else {
            ((x - min) as f64 / (max - min) as f64 * (bucket_count - 1) as f64).floor() as usize
        };
        buckets[idx].push(x);
    }

    for bucket in buckets.iter_mut() {
        insertion_sort(bucket);
    }

    let mut idx = 0;
    for bucket in buckets.iter() {
        for &x in bucket.iter() {
            a[idx] = x;
            idx += 1;
        }
    }
}

File: _includes/sort-benchmark/algorithms/bucket.swift
func bucket_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { bucket_sort($0) }
}

func bucket_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    if a.isEmpty {
        return
    }

    let n = a.count
    let min = a.min()!
    let max = a.max()!
    let bucket_count = n
    var buckets = Array(repeating: [Int](), count: bucket_count)

    for i in 0..<n {
        let x = a[i]
        let idx: Int
        if max == min {
            idx = 0
        } else {
            idx = Int((Double(x - min) / Double(max - min) * Double(bucket_count - 1)).rounded(.down))
        }
        buckets[idx].append(x)
    }

    for i in 0..<buckets.count {
        insertion_sort(&buckets[i])
    }

    var idx = 0
    for bucket in buckets {
        for x in bucket {
            a[idx] = x
            idx += 1
        }
    }
}
